bindings/go/ctf/tar.go

```diff
@@ -0,0 +1,272 @@
+	if flag&O_RDONLY != 0 || (flag&os.O_WRONLY == 0 && flag&os.O_RDWR == 0) {
+		ctf.FS().ForceReadOnly()
+	}
+
+	return ctf, nil
+}
+
+func extractTARToFilesystemCTF(reader *tar.Reader, ctf *FileSystemCTF) (err error) {
+	for {
+		header, err := reader.Next()
+		if errors.Is(err, io.EOF) {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		if strings.Contains(header.Name, "..") {
+			return fmt.Errorf("invalid tar entry, contains %q: %s", "..", header.Name)
+		}
+		switch header.Typeflag {
+		case tar.TypeReg:
+			if err := ctf.writeFile(header.Name, reader); err != nil {
+				return fmt.Errorf("unable to write file: %w", err)
+			}
+		case tar.TypeDir:
+			if err = os.MkdirAll(header.Name, 0o755); err != nil {
+				return fmt.Errorf("unable to create directory: %w", err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// Archive creates an archive from the provided CTF and writes it to the specified path.
+// The format of the archive is determined by the format parameter.
+// Supported formats are FormatTAR, FormatTGZ, and FormatDirectory.
+// If the format is FormatDirectory, the filesystem is copied to the specified path.
+func Archive(ctf CTF, path string, format FileFormat) error {
+	switch format {
+	case FormatDirectory:
+		return ArchiveDirectory(ctf, path)
+	case FormatTAR, FormatTGZ:
+		return ArchiveTAR(ctf, path, format)
+	default:
+		return ErrUnsupportedFormat
+	}
+}
+
+// ArchiveDirectory archives the CTF to the specified path with FormatDirectory.
+// The blobs are copied to the directory and the index is written to the index file.
+// The CTF is not modified and only read from.
+// The directory is created if it does not exist.
+// The blobs are written to the blobs directory concurrently.
+func ArchiveDirectory(ctf CTF, path string) error {
+	var fsCTF CTF
+
+	blobs, err := ctf.ListBlobs()
+	if err != nil {
+		return fmt.Errorf("unable to list blobs: %w", err)
+	}
+
+	if fsCTF, err = OpenCTFFromOSPath(path, O_RDWR|O_CREATE); err != nil {
+		return fmt.Errorf("unable to setup file system ctf: %w", err)
+	}
+	if err := os.Mkdir(path, 0o755); os.IsExist(err) {
+	} else if err != nil {
+		return fmt.Errorf("unable to create directory: %w", err)
+	} else {
+		if fsCTF, err = OpenCTFFromOSPath(path, O_RDWR); err != nil {
+			return fmt.Errorf("unable to setup file system ctf: %w", err)
+		}
+	}
+
+	if len(blobs) > 0 {
+		var wg sync.WaitGroup
+		wg.Add(len(blobs))
+		errs := make(chan error, len(blobs))
+
+		for _, digest := range blobs {
+			go func(digest string) {
+				defer wg.Done()
+				b, err := ctf.GetBlob(digest)
+				if err != nil {
+					errs <- fmt.Errorf("unable to get blob %s: %w", digest, err)
+					return
+				}
+				if err := fsCTF.SaveBlob(b); err != nil {
+					errs <- fmt.Errorf("unable to save blob %s: %w", digest, err)
+					return
+				}
+				errs <- nil
+			}(digest)
+		}
+		wg.Wait()
+
+		var err error
+		for i := 0; i < len(blobs); i++ {
+			err = errors.Join(<-errs)
+		}
+		if err != nil {
+			return err
+		}
```

Comment:
Use errgroup?

Reply:
Further, we should definitely add cancellations here, right? Via context.

Reply:
changed to errgroups, then noticed our io has no context cancellation, then added ctx to everything in CTF :D

Reply:
Yay. :D